src/plugins/projectexplorer/projectvcsstatus.cpp

```diff
@@ -0,0 +1,204 @@
+        for (auto iter = m_filePathNodeMap.constBegin(); iter != m_filePathNodeMap.constEnd(); ++iter) {
+            if (hasSameStartPath(folderPathOf(folderNode), iter.key()) && m_fileNodeFolderMap.value(iter.value()).contains(folderNode)) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool ProjectVcsStatus::updateProjectStatusCache()
+{
+    bool hasTrackedChanges = false;
+    bool hasUnTrackedChanges = false;
+    if (m_currentProject.count() != 0) {
+        auto & changeSet = m_projectStatusCache[m_currentProject];
+        auto * vcs = Core::VcsManager::findVersionControlForDirectory(Utils::FilePath::fromString(m_currentProject));
+        if (vcs != nullptr) {
+            auto [newTrackedChanges, newUntrackedChanges] = vcs->localChanges(Utils::FilePath::fromString(m_currentProject));
+            hasTrackedChanges |= newTrackedChanges != changeSet.trackedChanges;
+            if (hasTrackedChanges) {
+                changeSet.trackedChanges = newTrackedChanges;
+            }
+            hasUnTrackedChanges |= newUntrackedChanges != changeSet.untrackedChanges;
+            if (hasUnTrackedChanges) {
+                changeSet.untrackedChanges = newUntrackedChanges;
+            }
+        }
+    }
+    return hasTrackedChanges || hasUnTrackedChanges;
+}
+
+NodeFileList ProjectVcsStatus::changedFileNodesForCurrentProject() const
+{
+    auto * project = ProjectTree::currentProject();
+    if (project == nullptr || ! m_projectStatusCache.contains(project->projectDirectory().toString())) {
+        return {};
+    }
+    auto * projectNode = project->rootProjectNode();
+    if (projectNode == nullptr) {
+        return {};
+    }
+    QHash<QString, ProjectExplorer::FileNode *> fileMap;
+    const auto & projectStatusCache = m_projectStatusCache.value(project->projectDirectory().toString()).trackedChanges;
+    if (projectNode != nullptr) {
+        projectNode->forEachNode(
+            [&](FileNode * fileNode) {
+                if (fileNode != nullptr) {
+                    auto filePathString = fileNode->filePath().toString();
+                    auto result = projectStatusCache.find(filePathString);
+                    if (result != projectStatusCache.end()) {
+                        if (! fileMap.contains(filePathString)) {
+                            fileMap.insert(filePathString, fileNode);
+                        }
+                    }
+                }
+            },
+            {},
+            [&](const FolderNode * folderNode) -> bool {
+                if (folderNode == nullptr) {
+                    return false;
+                }
+                auto result = std::find_if(projectStatusCache.begin(),
+                                           projectStatusCache.end(),
+                                           [&](const QString & file) -> bool {
+                                               return hasSameStartPath(folderPathOf(folderNode), file);
```

Comment:
This feels kind of expensive `project_files*changeset*log(directories)` and brittle. Should we use a directory lookup here?

Reply:
The project tree has the big issue that it contains virtual nodes that has same folder paths on the same level, like:
```
foo[master]
   Source Files -->/foo/
     bar.cpp --> /foo/bar.cpp
   Header Files -->/foo/
     bar.h --> /foo/bar.h
```
When we have a change on `bar.h` and we only check the directory path (without `std::find_if(...)`), `Header Files` and `Source Files` would indicate changes.

Reply:
Yes, you are correct.

What I wanted to suggest is to speed up the lookup. The traversal of the virtual folders ist kept, but the lookup if the folder has a change can be sped up.